test/unit/template-parts.spec.js

```diff
@@ -0,0 +1,579 @@
+/* Adapted from https://github.com/dy/template-parts - ISC - Dmitry Iv. */
+
+import { assert } from '@open-wc/testing';
+import { TemplateInstance, AttrPart, AttrPartList } from '../../src/js/utils/template-parts.js';
+
+const test = it;
+const is = assert.deepEqual;
+const any = (a, b) => assert.include(b, a);
+
+test('create: applies data to templated text nodes', () => {
+  const template = document.createElement('template');
+  const originalHTML = `{{x}}`;
+  template.innerHTML = originalHTML;
+  const instance = new TemplateInstance(template, { x: 'Hello world' });
+  is(template.innerHTML, originalHTML);
+  const root = document.createElement('div');
+  root.appendChild(instance);
+  is(root.innerHTML, `Hello world`);
+});
+
+test('create: can render into partial text nodes', () => {
+  const template = document.createElement('template');
+  const originalHTML = `Hello {{x}}!`;
+  template.innerHTML = originalHTML;
+  const instance = new TemplateInstance(template, { x: 'world' });
+  is(template.innerHTML, originalHTML);
+  const root = document.createElement('div');
+  root.appendChild(instance);
+  is(root.innerHTML, `Hello world!`);
+});
+
+test('create: can render nested text nodes', () => {
+  const template = document.createElement('template');
+  const originalHTML = '<div><div>Hello {{x}}!</div></div>';
+  template.innerHTML = originalHTML;
+  const instance = new TemplateInstance(template, { x: 'world' });
+  is(template.innerHTML, originalHTML);
+  const root = document.createElement('div');
+  root.appendChild(instance);
+  is(root.innerHTML, `<div><div>Hello world!</div></div>`);
+});
+
+test('create: applies data to templated attributes', () => {
+  const template = document.createElement('template');
+  const originalHTML = `<div class="{{y}}"></div>`;
+  template.innerHTML = originalHTML;
+  const instance = new TemplateInstance(template, { y: 'foo' });
+  is(template.innerHTML, originalHTML);
+  const root = document.createElement('div');
+  root.appendChild(instance);
+  is(root.innerHTML, `<div class="foo"></div>`);
+});
+
+test('create: can render into partial attribute nodes', () => {
+  const template = document.createElement('template');
+  const originalHTML = `<div class="my-{{y}}-state"></div>`;
+  template.innerHTML = originalHTML;
+  const instance = new TemplateInstance(template, { y: 'foo' });
+  is(template.innerHTML, originalHTML);
+  const root = document.createElement('div');
+  root.appendChild(instance);
+  is(root.innerHTML, `<div class="my-foo-state"></div>`);
+});
+
+test('create: can render into many values', () => {
```

Comment:
suggestion(non-blocking): Add tests for "partial args" and "no args", but for the constructor, similar to `update` tests. Aka, using this test as an example, how do these behave:

- `new TemplateInstance(template, { x: 'foo' })`
- `new TemplateInstance(template)`